File: lab5-demos/retrofit-demo-java/app/src/main/java/com/smd/retrofitdemo/MainActivity.java
package com.smd.retrofitdemo;

import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.TextView;
import java.util.ArrayList;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.RecyclerView;

/**
 * Shows a list of the astronauts who are on ISS or in transit to it
 */
public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getName();
    private PeopleInSpaceViewModel peopleInSpaceViewModel;
    private PeopleInSpaceAdapter peopleInSpaceAdapter;
    private TextView peopleCountTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initViewModel();

        initUI();
    }

    private void initViewModel() {
        peopleInSpaceViewModel = ViewModelProviders.of(this).get(PeopleInSpaceViewModel.class);
        peopleInSpaceViewModel.init();
        peopleInSpaceViewModel.getPeopleInSpaceLiveData().observe(this, peopleInSpace -> {
            if (peopleInSpace != null) {
                Log.d(TAG, "people in space " + peopleInSpace.getNumber());

                // Update the recycler view adapter
                peopleInSpaceAdapter.replaceAll(peopleInSpace.getPeople());

                peopleCountTextView.setText(getString(R.string.astronauts_count, peopleInSpace.getNumber()));
            }
        });
    }

    private void initUI() {
        peopleInSpaceAdapter = new PeopleInSpaceAdapter(new ArrayList<>());

        Button showPeopleButton = findViewById(R.id.show_button);
        showPeopleButton.setOnClickListener(v -> peopleInSpaceViewModel.getPeopleInSpace());

        RecyclerView astronauts = findViewById(R.id.astronauts_list);
        astronauts.setAdapter(peopleInSpaceAdapter);

        peopleCountTextView = findViewById(R.id.count_tv);
    }
}

File: lab5-demos/retrofit-demo-kotlin/app/src/main/java/com/smd/retrofitdemo/MainActivity.kt
package com.smd.retrofitdemo

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProviders
import com.smd.retrofitdemo.databinding.ActivityMainBinding
import com.smd.retrofitdemo.network.PeopleInSpace
import timber.log.Timber
import java.util.ArrayList

/**
 * Shows a list of the astronauts who are on ISS or in transit to it
 */
class MainActivity : AppCompatActivity() {
    private lateinit var peopleInSpaceViewModel: PeopleInSpaceViewModel
    private lateinit var peopleInSpaceAdapter: PeopleInSpaceAdapter
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // view binding (the ActivityMainBinding is generated based on the layout file's name)
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)

        initViewModel()
        initUI()
    }

    private fun initViewModel() {
        peopleInSpaceViewModel = ViewModelProviders.of(this).get(
            PeopleInSpaceViewModel::class.java
        )
        peopleInSpaceViewModel.init()
        peopleInSpaceViewModel.peopleInSpaceLiveData?.observe(
            this,
            { peopleInSpace: PeopleInSpace? ->
                if (peopleInSpace != null) {
                    Timber.d("People in space ${peopleInSpace.number}")

                    // Update the recycler view adapter
                    peopleInSpaceAdapter.replaceAll(peopleInSpace.people)
                    binding.astronautsCountTextview.text =
                        getString(R.string.astronauts_count, peopleInSpace.number)
                }
            })
    }

    private fun initUI() {
        peopleInSpaceAdapter = PeopleInSpaceAdapter(ArrayList())

        binding.showButton.setOnClickListener { peopleInSpaceViewModel.peopleInSpace }
        binding.astronautsList.adapter = peopleInSpaceAdapter
    }
}